### PR TITLE
Use CopyOnWriteArrayList instead of recreating manually

### PR DESCRIPTION
**Pull request by kezhenxu94**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

https://github.com/apache/skywalking/blob/8faa3564037570fca198ab9f70534f3d359a9d40/apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/consumer/MultipleChannelsConsumer.java#L99-L109

the comment `// Recreate the new list to avoid change list while the list is used in consuming.` reminds me of the `CopyOnWrite` data structure that is perfectly suitable to this scenario, although the method `addNewTarget` is not frequently invoked, I did a simple benchmark out of curiosity.

This patch also fixes that the non-atomic operation `+=` on the `volatile` variable `size`

### CODE REVIEW

File: apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/CowVsManualCopyBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        Options opt = new OptionsBuilder()
+            .include(CowVsManualCopyBenchmark.class.getSimpleName())
+            .mode(Mode.Throughput)
+            .forks(1)
+            .warmupIterations(4)
+            .measurementIterations(4)
+            .build();
+
+        new Runner(opt).run();
+    }
```
Comment (wu-sheng) [inline apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/CowVsManualCopyBenchmark.java]: You should post benchmark result here.

**Comment by kezhenxu94** [comment]
FYI @wu-sheng , I'm recently considering whether we should create a new module/directory to place all the benchmark tests like other projects, the reasons are:

1. we have some benchmarks that are independent to any existed module, such as `LinkedArrayBenchmark`, `StringFormatBenchmarkTest`, IMHO they should be placed in an independent module, otherwise it's hard for new-comers to refer to and find out whether we already have that benchmark or not

2. some of the existed benchmarks are annotated with `@Test`, and are executed in maven test phase (I think), which is not necessary and time consuming, gathering them together may be easy to manage and remove `@Test`

**Comment by wu-sheng** [comment]
@kezhenxu94 How about we create WIKI pages for that? And link to the PRs?

**Comment by wu-sheng** [comment]
We have this. https://cwiki.apache.org/confluence/display/SKYWALKING/Home 

I could grant committers access right. Then all the acceptable performance improvement PRs, should have a benchmark list there.

**Comment by kezhenxu94** [comment]
> @kezhenxu94 How about we create WIKI pages for that? And link to the PRs?

That's a great alternative, it would be easy for new-comers who are interested in performance optimization and can refer to the wiki

**Comment by dmsolr** [comment]
Out of this patch.
ArrayList has worse performance because ArrayList uses `add()` one by one. It would be better if using `addAll()` or `new ArrayList<>(collections)`.

**Comment by wu-sheng** [comment]
We accept a lot of performance improvements these days, but we don't need to focus on that too much.

**Comment by wu-sheng** [comment]
Like I said in some other PRs, we don't require people to change their coding style, due to that is personal interests.

It is only acceptable when causing runtime performance issue.

**Comment by dmsolr** [comment]
Yes. In this case, it is more importance to fix `size`. And I say that it is out of this issue. Just for talking.

**Pull request closed without merging**